X2WOTCCommunityPromotionScreen/Src/X2WOTCCommunityPromotionScreen/Classes/CPS_UIArmory_PromotionHero.uc

```diff
@@ -466,7 +466,7 @@ function bool UpdateAbilityIcons_Override(out CPS_UIArmory_PromotionHeroColumn C
 				// Look ahead to the next rank and check to see if the current ability is a prereq for the next one
 				// If so, turn on the connection arrow between them
 				//if (Column.Rank < (class'X2ExperienceConfig'.static.GetMaxRank() - 2) && Unit.GetRank() > (Column.Rank + 1))
-				if (Column.Rank < (Columns.Length - 2) && Unit.GetRank() > (Column.Rank + 1) || `GETMCMVAR(SHOW_UNREACHED_PERKS)) // Issue #61 - always connect abilities if "Show unreached perks" is enabled.
+				if (Column.Rank < (Columns.Length - 2) && (Unit.GetRank() > (Column.Rank + 1) || `GETMCMVAR(SHOW_UNREACHED_PERKS))) // Issue #61 - always connect abilities if "Show unreached perks" is enabled.
```

Comment:
Fixed an incorrect logical expression, which caused the CPS to attempt to process abilities from non-existent ranks if the "show unreached perks" option was enabled.

Reply:
Why is this a rank check against `Columns.Length - 2`? Other than it's the original code. Wouldn't this fail if a Colonel ability was a prerequisite of a Brigadier one? Sorry, my brain's still not functioning very well, so may be missing something obvious.

Reply:
`Column.Rank` is the soldier rank at which the column is located. It is indexed from 0.
If the `Column.Rank` is 0, it means this is the column for the first soldier's rank.

`Columns.Length` is the overall number of columns for this soldier.

So let's say the `Columns.Length` is `6`, so the soldier has six ranks total, so the `Column.Rank` will take values from 0 to 5.

The `Column.Rank < (Columns.Length - 1)` can be translated as "is 0...5 lower than 6", which will always succeed, and is essentially "is the current column is inside the array of all columns for this soldier".

The `Column.Rank < (Columns.Length - 2)` can be translated as "is 0...5 lower than 5", which will fail only for the final column, the column for the final rank available for this soldier. Essentially the check is "is this column not the final one".

Reply:
> The Column.Rank < (Columns.Length - 1) can be translated as "is 0...5 lower than 6"

No, that's translated as "is 0..5 lower than 5. (Columns.Length - 1) is 5, not 6.

Reply:
Oh damn, you're right. I wonder if the original code is bugged.

Reply:
Well, since this is the original behavior, I suggest we merge this as is and check the functionality of this condition in another issue.